dipy/reconst/cti.py

```diff
@@ -0,0 +1,658 @@
+
+import warnings
+import functools
+import numpy as np
+import scipy.optimize as opt
+from dipy.core.optimize import PositiveDefiniteLeastSquares
+from dipy.reconst.multi_voxel import multi_voxel_fit
+
+from dipy.reconst.base import ReconstModel
+from dipy.reconst.utils import cti_design_matrix as design_matrix
+from dipy.reconst.dki import (
+    DiffusionKurtosisFit,
+    apparent_kurtosis_coef,
+    mean_kurtosis,
+    axial_kurtosis
+    )
+from dipy.data import load_sdp_constraints
+from dipy.reconst.dti import (
+    decompose_tensor, from_lower_triangular, lower_triangular, mean_diffusivity, MIN_POSITIVE_SIGNAL)
+from dipy.reconst.qti import from_6x1_to_3x3
+# sources of kurtosis:
+# we've formulats for Kaniso, Ksio, then get Ktotal. So for microscopic kurtosis, we subtract, Kt - K aniso
+from dipy.core.onetime import auto_attr
+
+
+
+def split_cti_params(cti_params): #here cti_params.shape : (48, )
+    r"""Extract the diffusion tensor eigenvalues, the diffusion tensor eigenvector matrix, and the 21 independent elements of the covariance tensor, and the 15 independent elements of the kurtosis tensor from the model parameters estimated from the CTI model
+    Parameters:
+         -----------
+         params: numpy.ndarray (...,48)
+                 All parameters estimated from the correlation tensor model.
+                 Paramters are ordered as follows::
+
+                 1. Three diffusion tensor's eigenvalues
+                 2. Three lines of the eigenvector matrix each containing the
+                 first, second and third coordinates of the eigenvector
+                 3. Fifteen elements of the kurtosis tensor
+                 4. Twenty-One elements of the covariance tensor
+         S0 : float or ndarray (optional)
+             The non diffusion-weighted signal in every voxel, or across all
+             voxels. Default: 100
+
+         Returns
+         -------
+         evals: Three diffusion tensor's eigenvalues
+         evecs: Three lines of the eigenvector matrix each continaint ehf irst, second and third coordinates of the eigenvector
+         kt: Fifteen elemnets of the kurtosis tensor
+         cvt : Twenty-one elements of the covariance tensor
+
+       """
+    # DT_elements = np.squeeze(cti_params[:6, ...])
+    # evals, evecs = decompose_tensor(from_lower_triangular(DT_elements))
+    evals = cti_params[..., :3]
+    evecs = cti_params[..., 3:12].reshape(cti_params.shape[:-1] + (3, 3))
+    # kt = cti_params[12:27, ...] #original
+    kt = cti_params[..., 12:27]
+    # cvt = cti_params[27:48, ...]
+    cvt = cti_params[..., 27:48]
+    return evals, evecs, kt, cvt
+
+
+def cti_prediction(cti_params, gtab1, gtab2, S0=100):
+    """Predict a signal given correlation tensor imaging parameters
+
+        Parameters
+        ----------
+        cti_params: numpy.ndarray (...,48)
+                All parameters estimated from the correlation tensor model.
```

Comment:
are you sure that's ok because in DKI, we've: 
dki_params : ndarray (x, y, z, 27) or (n, 27)
        All parameters estimated from the diffusion kurtosis model.
        Parameters are ordered as follows:
            1) Three diffusion tensor's eigenvalues
            2) Three lines of the eigenvector matrix each containing the first,
               second and third coordinates of the eigenvectors respectively
            3) Fifteen elements of the kurtosis tensor